lib/iris/fileformats/rules.py

```diff
@@ -859,6 +859,56 @@ def _make_cube(field, converter):
     return cube, metadata.factories, metadata.references
 
 
+def _resolve_references(results_needing_reference, concrete_reference_targets,
+                        filenames=''):
+    regrid_cache = {}
+    for cube, factories in results_needing_reference:
+        for factory in factories:
+            try:
+                args = _dereference_args(factory, concrete_reference_targets,
+                                         regrid_cache, cube)
+            except _ReferenceError as e:
+                msg = 'Unable to create instance of {factory}. ' + str(e)
+                factory_name = factory.factory_class.__name__
+                warnings.warn(msg.format(filenames=filenames,
+                                         factory=factory_name))
+            else:
+                aux_factory = factory.factory_class(*args)
+                cube.add_aux_factory(aux_factory)
+        yield cube
+
+
+def as_cubes(fields, converter, callback=None):
+    concrete_reference_targets = {}
+    results_needing_reference = []
+
+    for field in fields:
+        # Convert the field to a Cube.
+        cube, factories, references = _make_cube(field, converter)
+
+        cube = iris.io.run_callback(user_callback, cube, field, filename)
+
+        # Cross referencing
+        for reference in references:
+            name = reference.name
+            # Register this cube as a source cube for the named
+            # reference.
+            target = concrete_reference_targets.get(name)
+            if target is None:
+                target = ConcreteReferenceTarget(name, reference.transform)
+                concrete_reference_targets[name] = target
+            target.add_cube(cube)
+
+        if factories:
+            results_needing_reference.append((cube, factories))
+        else:
+            yield cube
+
+    for cube in _resolve_references(results_needing_reference,
+                                    concrete_reference_targets):
+        yield cube
+
+
 def load_cubes(filenames, user_callback, loader, filter_function=None):
```

Comment:
consideration has been given to this.  A clean factoring out has not been evident.  There is some duplication of logic in this implementation, but it is limited and the processing is not identical, so care has to be taken which ever way round we go.

I'm happy to take direction, but I haven't seen a neater solution than this in my travels; they have been meandering and I have seen significantly less neat implementations than this one